Description: Fix ftbfs with GCC-10

Author: Sudip Mukherjee <sudipm.mukherjee@gmail.com>
Bug-Debian: https://bugs.debian.org/957582
Forwarded: no

---

--- ncftp-3.2.5.orig/sh_util/gpshare.c
+++ ncftp-3.2.5/sh_util/gpshare.c
@@ -28,7 +28,7 @@
 
 static int gIsAtty1 = 1, gIsAtty2 = 1;
 extern int gLoadedBm, gBookmarkMatchMode;
-Bookmark gBm;
+extern Bookmark gBm;
 
 double
 FileSize(double size, const char **uStr0, double *uMult0)
