--- grub-2.06/util/grub-mkconfig.in.orig	2021-03-10 10:42:22.000000000 -0300
+++ grub-2.06/util/grub-mkconfig.in	2021-09-13 16:48:03.060695839 -0300
@@ -141,7 +141,7 @@
 GRUB_DEVICE_BOOT_UUID="`${grub_probe} --device ${GRUB_DEVICE_BOOT} --target=fs_uuid 2> /dev/null`" || true
 
 # Disable os-prober by default due to security reasons.
-GRUB_DISABLE_OS_PROBER="true"
+GRUB_DISABLE_OS_PROBER="false"
 
 # Filesystem for the device containing our userland.  Used for stuff like
 # choosing Hurd filesystem module.
