--- grub-2.06/util/grub.d/10_linux.in.orig	2020-12-22 20:24:17.000000000 -0300
+++ grub-2.06/util/grub.d/10_linux.in	2021-09-09 09:23:53.809125172 -0300
@@ -218,7 +218,8 @@
 	   "initramfs-genkernel-${version}" \
 	   "initramfs-genkernel-${alt_version}" \
 	   "initramfs-genkernel-${GENKERNEL_ARCH}-${version}" \
-	   "initramfs-genkernel-${GENKERNEL_ARCH}-${alt_version}"; do
+	   "initramfs-genkernel-${GENKERNEL_ARCH}-${alt_version}" \
+	   "initrd.gz" "initrd.lz"; do
     if test -e "${dirname}/${i}" ; then
       initrd_real="${i}"
       break
